<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Motif analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Explore eQTL</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/explore-eqtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Motif analysis</h1>
<h4 class="date"><em>22 June, 2018</em></h4>

</div>


<div id="set-up" class="section level1">
<h1><span class="header-section-number">1</span> Set up</h1>
<pre class="r"><code>source(&#39;../scripts/mylib.R&#39;)
library(ggplot2)
library(dplyr)
library(BSgenome.Hsapiens.UCSC.hg19)
library(seqinr)
library(mashr)
set.seed(2018)</code></pre>
</div>
<div id="read-data" class="section level1">
<h1><span class="header-section-number">2</span> Read data</h1>
<p>In this section, the motif analysis on the promoters of the genes with strong loading on “heart factor” is performed. The motivation is to see whether “tissue-specific” (strong loading on “heart factor”) eQTL is driven by tissue-specific promoter.</p>
<pre class="r"><code>flash &lt;- readRDS(&#39;../output/gtex_flash.rds&#39;)
eqtl_annotated &lt;- read.table(&#39;../output/strong_eqtl_annotated.txt.gz&#39;, header = T)
# eqtl_annotated.dup &lt;- duplicated(eqtl_annotated$ensembl_gene_id) 
# eqtl_annotated &lt;- eqtl_annotated[!eqtl_annotated.dup, ]
# eqtl_annotated$start.merged &lt;- e$s
# eqtl_annotated$end.merged &lt;- e$e
mash &lt;- readRDS(&#39;../output/gtex_mash.rds&#39;)
eqtl.names &lt;- read.table(&#39;../output/strong_eqtl_names.txt&#39;)</code></pre>
</div>
<div id="obtain-positive-and-negative-eqtls" class="section level1">
<h1><span class="header-section-number">3</span> Obtain positive and negative eQTLs</h1>
<p>The idea is to select eQTLs that have clear tissue-specific pattern, <em>i.e.</em> it has strong effect in heart/muscle but no strong effect on others.</p>
<pre class="r"><code>heart.idx &lt;- c(25, 26, 29)
cond1 &lt;- abs(l.five) &gt; 8
mash.mean &lt;- mash$posterior$result$PosteriorMean
o.mean &lt;- rowMeans(mash.mean[, !(1 : 44 %in% heart.idx)])
h.mean &lt;- rowMax(mash.mean[, (1 : 44 %in% heart.idx)])
cond2 &lt;- abs(o.mean) &lt; abs(h.mean) * 0.2
mash.heart.ind &lt;- cond2 &amp; cond1
sum(mash.heart.ind)</code></pre>
<pre><code>## [1] 42</code></pre>
<pre class="r"><code>e &lt;- seq(1 : length(mash.heart.ind))[mash.heart.ind]
npart &lt;- ceiling(length(e) / 4)
for(n in 1 : npart) {
  par(mfrow = c(2, 2))
  for(i in ((n - 1) * 4 + 1) : min(length(e), (n * 4)))  {
    mash_plot_meta(mash$posterior, e[i], color = meta.colors(lines = tissue.color), main = paste0(e[i], &#39;: &#39;, eqtl[i]))
  }
}</code></pre>
<p><img src="motif_analysis_files/figure-html/unnamed-chunk-482-1.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-2.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-3.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-4.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-5.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-6.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-7.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-8.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-9.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-10.png" width="768" /><img src="motif_analysis_files/figure-html/unnamed-chunk-482-11.png" width="768" /></p>
<!-- # Extracting promoter -->
<!-- Promoter is defined as the 1 kb region upstream of TSS. -->
<!-- ```{r} -->
<!-- window.size <- 1000 -->
<!-- l.five <- flash$EL[, 5] -->
<!-- l.one <- flash$EL[, 1] -->
<!-- # l.five <- l.five[eqtl_annotated$idx] -->
<!-- # l.one <- l.one[eqtl_annotated$idx] -->
<!-- eqtl.names$loading_5 <- l.five -->
<!-- eqtl.names$loading_1 <- l.one -->
<!-- eqtl.names$idx <- 1 : nrow(eqtl.names) -->
<!-- eqtl.names.sub <- eqtl.names[(abs(eqtl.names$loading_5) > 5 & abs(eqtl.names$loading_1) < 1) | (abs(eqtl.names$loading_5) < 1 & abs(eqtl.names$loading_1) > 50), ] -->
<!-- # df.sub <- eqtl_annotated[(abs(eqtl_annotated$loading_5) > 5 & abs(eqtl_annotated$loading_1) < 1) | (abs(eqtl_annotated$loading_5) < 1 & abs(eqtl_annotated$loading_1) > 50), ] -->
<!-- df.sub <- eqtl_annotated[eqtl_annotated$idx %in% eqtl.names.sub$idx, ] -->
<!-- df.sub$specific <- TRUE -->
<!-- df.sub$specific[df.sub$idx %in% eqtl.names$idx[(abs(eqtl.names$loading_5) < 1 & abs(eqtl.names$loading_1) > 50)]] <- FALSE -->
<!-- df.sub$tss <- df.sub$transcript_start -->
<!-- df.sub$tss[df.sub$strand == -1] <- df.sub$transcript_end[df.sub$strand == -1] -->
<!-- tss_up1kb <- df.sub$tss - df.sub$strand * window.size -->
<!-- df.sub$tss_big <- df.sub$tss -->
<!-- df.sub$tss_big[df.sub$strand == -1] <- tss_up1kb[df.sub$strand == -1] -->
<!-- df.sub$tss_small <- tss_up1kb -->
<!-- df.sub$tss_small[df.sub$strand == -1] <- df.sub$tss[df.sub$strand == -1] -->
<!-- e <- df.sub %>% group_by(ensembl_gene_id) %>% summarise(union = getUnion(tss_small, tss_big), idx = unique(idx)) -->
<!-- e <- e[order(e$idx), ] -->
<!-- df.sub <- df.sub[order(df.sub$idx), ] -->
<!-- df.sub <- df.sub[!duplicated(df.sub$ensembl_gene_id), ] -->
<!-- df.sub[, 'union'] <- e[, 'union'] -->
<!-- df.sub <- expendUnion(df.sub) -->
<!-- df.sub$seq <- getFasta(data.frame(paste0('chr', df.sub$chromosome_name), df.sub$interval.start, df.sub$interval.end), BSgenome.Hsapiens.UCSC.hg19) -->
<!-- df.sub.pos <- df.sub[df.sub$specific, ] -->
<!-- write.fasta(as.list(df.sub.pos$seq), paste0(df.sub.pos$idx, ':', df.sub.pos$chromosome_name, ':', df.sub.pos$interval.start, '-', df.sub.pos$interval.end), file.out = '../output/motif_analysis_factor_five.positive.fa') -->
<!-- df.sub.neg <- df.sub[!df.sub$specific, ] -->
<!-- write.fasta(as.list(df.sub.neg$seq), paste0(df.sub.neg$idx, ':', df.sub.neg$chromosome_name, ':', df.sub.neg$interval.start, '-', df.sub.neg$interval.end), file.out = '../output/motif_analysis_factor_five.negative.fa') -->
<!-- ``` -->
<!-- # MEME Suite result (AME) -->
<!-- AME performs enrichment analysis by comparing the occurence of motif between positive sequence set and control sequence set. Here, the motifs are from motif database, HOCOMOCOv11 full HUMAN. See output file [here](https://htmlpreview.github.io/?https://github.com/liangyy/explore-eqtl/blob/master/output/AME_f5.htm) -->
<!-- ## TFDP1 -->
<!-- The relation of TFDP1 and heart/muscle tissue is not found but TFDP2 shows high expression level in muscle and heart tissue. Also, TFDP2 is involved in heart development (related to cell cycle).   -->
<!-- ## EGR2 -->
<!-- No evidence has been found. -->
<!-- ## SP1  -->
<!-- [@flesch2001trail] reported that sp1 is involved in the tissue-specific activation of promoter. -->
<!-- ## EGR1 -->
<!-- Many studies have shown that EGR1 plays a role in cardic myocyte. -->
<!-- # MEME Suite result (FIMO) -->
<!-- FIMO scans and scores the sequences by given motif. Here, sp1 is used (JASPAR ID MA0079.1). -->
<!-- ```{r} -->
<!-- pos <- read.table('../output/fimo_f5.positive.tsv', header = T) -->
<!-- neg <- read.table('../output/fimo_f5.negative.tsv', header = T) -->
<!-- pos.pmin <- pos %>% group_by(sequence_name) %>% -->
<!--   summarise(score = min(p.value)) -->
<!-- neg.pmin <- neg %>% group_by(sequence_name) %>% -->
<!--   summarise(score = min(p.value)) -->
<!-- df.sub.order <- df.sub[order(df.sub$idx, decreasing = F), ] -->
<!-- pos.pmin <- pos.pmin[order(pos.pmin$sequence_name, decreasing = F), ] -->
<!-- neg.pmin <- neg.pmin[order(neg.pmin$sequence_name, decreasing = F), ] -->
<!-- df.sub.order$pmin <- NA  -->
<!-- df.sub.order$pmin[df.sub.order$idx %in% pos.pmin$sequence_name] <- pos.pmin$score -->
<!-- df.sub.order$pmin[df.sub.order$idx %in% neg.pmin$sequence_name] <- neg.pmin$score -->
<!-- ggplot(df.sub.order) + geom_boxplot(aes(x = specific, y = pmin, color = specific)) + scale_y_log10() + ggtitle('Q-value vs. loading on factor 5') + geom_jitter(aes(x = specific, y = pmin, color = specific), height = 0) -->
<!-- wilcox.test(df.sub.order$pmin[df.sub.order$specific], df.sub.order$pmin[!df.sub.order$specific], alternative = 'l') -->
<!-- ggplot(df.sub.order) + geom_bar(aes(group = is.na(pmin), x = specific, fill = is.na(pmin)), position = "fill") + ggtitle('Fraction of significant hits in positive and control sets') + labs(fill = 'Motif not show', x = 'Control = FALSE, positive = TRUE') -->
<!-- ``` -->
<!-- The result indicates that the promoter region of factor 5 specific genes have significantly more sp1 binding sites than general eQTLs (strong loading on 1 but weak loading on 5). So, it seems that the tissue-specific eQTLs for factor 5 are indeed controlled by tissue-specific TFs but it is still questionable about how it is regulated (by promoter or chromatin looping mediated promoter-enhancer interaction). The distance analysis gives no pattern (see [here](position_to_enhancer.html)).  -->
<!-- # Save data -->
<!-- ```{r} -->
<!-- saveRDS(df.sub.order, '../output/motif_analysis_FIMO.rds') -->
<!-- ``` -->
</div>
<div id="references" class="section level1">
<h1><span class="header-section-number">4</span> References</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
